datafusion/expr-common/src/scalar.rs

```diff
@@ -0,0 +1,146 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::{
+    array::{Array, ArrayRef},
+    datatypes::DataType,
+};
+use datafusion_common::{exec_err, DataFusionError, Result, ScalarValue};
+
+/// Represents a physical scalar value obtained after evaluating an expression.
+///
+/// Currently [`Scalar`] represents a no-op wrapper around scalar value. This
+/// is due to the fact that the current variants of [`ScalarValue`] correspond
+/// 1:1 with the variants of [`DataType`]. The goal of this type is aid the
+/// process of condensing some of the variants (e.g. [`ScalarValue::LargeUtf8`],
+/// [`ScalarValue::Utf8View`], and [`ScalarValue::Utf8`]) into a single variant
+/// (e.g. [`ScalarValue::Utf8`]). This type achieves this goal by storing the
+/// physical type as [`DataType`] alongside the value represented as
+/// [`ScalarValue`].
+///
+/// [`Scalar`] cannot be constructed directly as not all [`DataType`] variants
```

Comment:
I don't think this is intentional (that some types can not be handled)

Reply:
I'm not sure what you exactly mean. Currently (but potentially not in the future) the `DataType` of `Scalar` should be at least logically equivalent to the data type reported by `ScalarValue::data_type()`